### PR TITLE
Handling querypb.BindVariable better for lists.

### PR DESCRIPTION
**Pull request by alainjobart**
Handling list bind variables type *querypb.BindVariable with Type TUPLE, in vtgate and vttablet.

### CODE REVIEW

**Comment by alainjobart** [comment]
@sougou I am not sure the vttablet side is correct, there is a FIXME comment in there, input appreciated. Thanks!

File: go/vt/tabletserver/codex.go [hunk 1]
```diff
@@ -96,19 +97,43 @@ func resolveListArg(col *schema.TableColumn, key string, bindVars map[string]int
...
+	case *querypb.BindVariable:
+		if list.Type != querypb.Type_TUPLE {
+			return nil, NewTabletError(vtrpcpb.ErrorCode_BAD_INPUT, "expecting list for bind var %s: %v", key, list)
+		}
+		resolved := make([]sqltypes.Value, len(list.Values))
+		for i, v := range list.Values {
+			// FIXME(alainjobart,sougou): we ignore the
+			// input's v.Type in this expression. I am not
+			// sure this is the way to go.
+			sqlval, err := sqltypes.BuildConverted(col.Type, v.Value)
```
Comment (sougou) [inline go/vt/tabletserver/codex.go]: All you have to do is pass this in instead: `sqltypes.MaleTrusted(v.Typ, v.Value)`.
Then BuildConverted will `do the right thing`.

File: go/vt/vtgate/router.go [hunk 2]
```diff
@@ -418,24 +418,41 @@ func (rtr *Router) execInsertSharded(vcursor *requestContext, route *engine.Rout
...
+		case []interface{}:
+			return l, nil
+		case *querypb.BindVariable:
+			if l.Type != querypb.Type_TUPLE {
+				return nil, fmt.Errorf("expecting list for bind var %s: %v", v, list)
+			}
+			result := make([]interface{}, len(l.Values))
+			for i, val := range l.Values {
+				var err error
+				result[i], err = sqltypes.ValueFromBytes(val.Type, val.Value)
```
Comment (sougou) [inline go/vt/vtgate/router.go]: If these values are assumed correct, we could use the more efficient `sqltypes.MakeTrusted`.

File: go/vt/tabletserver/codex.go (hunk 1 above)
Comment (alainjobart) [inline go/vt/tabletserver/codex.go]: I think this is the first time we use a bind variable here, from the client, so we have to make sure its type matches its content. I don't think calling MakeTrusted is right.

The problem I have here is:
- we have the bind variable type + value
- we have the column type
You're saying I should build the sqltypes.Value from bind variable type + value, and then call validateValue on the result?

File: go/vt/vtgate/router.go (hunk 2 above)
Comment (alainjobart) [inline go/vt/vtgate/router.go]: Are they assumed correct? I think it's the first time we use them, we should check them?

File: go/vt/tabletserver/codex.go (hunk 1 above)
Comment (sougou) [inline go/vt/tabletserver/codex.go]: The idea is that you validate the bind var when you build it. Then you assume it's correct. If so, you use `MakeTrusted`. If you don't trust the values, then you should use ValueFromBytes.

I think it's safe to assume that BindVars are validated at build time, right?

**Comment by alainjobart** [comment]
@sougou PTAL I think this is what needs to be done, thanks for the feedback!

**Comment by sougou** [comment]
LGTM

**Pull request merged**